cmd/kube-apiserver/app/testing/testserver.go

```diff
@@ -37,14 +37,15 @@ import (
 // TearDownFunc is to be called to tear down a test server.
 type TearDownFunc func()
 
-// StartTestServer starts a etcd server and kube-apiserver. A rest client config and a tear-down func
-// are returned.
+// StartTestServer starts a etcd server and kube-apiserver. A rest client config and a tear-down func,
+// and location of the tmpdir are returned.
 //
 // Note: we return a tear-down func instead of a stop channel because the later will leak temporariy
 // 		 files that becaues Golang testing's call to os.Exit will not give a stop channel go routine
 // 		 enough time to remove temporariy files.
-func StartTestServer(t *testing.T) (result *restclient.Config, tearDownForCaller TearDownFunc, err error) {
-	var tmpDir string
+//
+func StartTestServer(t *testing.T, saPubKeyPath string, plugins []string, insecure bool) (result *restclient.Config, tearDownForCaller TearDownFunc, tmpDir string, err error) {
```

Comment:
Can we make a `TestServer struct` out of all the return values?

Reply:
Please rebase onto https://github.com/kubernetes/kubernetes/pull/51685. I added a `flags []string` to control the config. I don't want to get any more logic in here in this func. Flags enforce that we mostly have an e2e setup as with a real kube-apiserver.

Reply:
Okay i guess you mean i wait for the PR to be merged? before continuing to work on this?  If i rebase now wont this PR also get all the commits #51685 as its still not in the master yet? (Apologies for the silly question i have not rebased from a PR to a PR before, just confused)

How about a struct with methods? 
```
type TestServer struct {
         //All the return values
}

func (T *TestServer) Start(....)....
func (T *TestServer) StartTestServerOrDie(...)...
.
.
.
```
Okay? 

If yes , I will update this PR to only make this pkg from `set of functions`  to `struct with methods` along with updating all its invocations.

Reply:
just the return values, not the inputs, i.e. also no methods. 

IMO we should not add anything to the inputs anymore, but use the flags instead (as done in my PR). Everything else results in non-standard apiserver instances again which this whole machanism was trying to fix (compared to the situation we have in integration tests).